imap.go

```diff
@@ -1,102 +1,182 @@
 package maddy
 
 import (
-	"io"
-	"log"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"net"
 	"os"
+	"strings"
+	"sync"
 
-	imapcompress "github.com/emersion/go-imap-compress"
-	imapproxy "github.com/emersion/go-imap-proxy"
 	imapbackend "github.com/emersion/go-imap/backend"
 	imapserver "github.com/emersion/go-imap/server"
-	imappgp "github.com/emersion/go-pgpmail/imap"
-	pgplocal "github.com/emersion/go-pgpmail/local"
-	"github.com/mholt/caddy/caddyfile"
+	"github.com/emersion/maddy/config"
+	"github.com/emersion/maddy/log"
+	"github.com/emersion/maddy/module"
+
+	appendlimit "github.com/emersion/go-imap-appendlimit"
+	compress "github.com/emersion/go-imap-compress"
+	move "github.com/emersion/go-imap-move"
+	"github.com/foxcpp/go-sqlmail/imap/children"
 )
 
-func newIMAPServer(tokens map[string][]caddyfile.Token) (server, error) {
-	var be imapbackend.Backend
-	if tokens, ok := tokens["proxy"]; ok {
-		var err error
-		be, err = newIMAPProxy(caddyfile.NewDispenserTokens("", tokens))
-		if err != nil {
-			return nil, err
-		}
+type IMAPEndpoint struct {
+	name      string
+	serv      *imapserver.Server
+	listeners []net.Listener
+	Auth      module.AuthProvider
+	Store     module.Storage
+
+	tlsConfig   *tls.Config
+	listenersWg sync.WaitGroup
+
+	Log log.Logger
+}
+
+func NewIMAPEndpoint(_, instName string) (module.Module, error) {
+	endp := &IMAPEndpoint{
+		name: instName,
+		Log:  log.Logger{Out: log.StderrLog, Name: "imap"},
 	}
+	endp.name = instName
 
-	if tokens, ok := tokens["pgp"]; ok {
-		var err error
-		be, err = newIMAPPGP(caddyfile.NewDispenserTokens("", tokens), be)
+	return endp, nil
+}
+
+func (endp *IMAPEndpoint) Init(globalCfg map[string]config.Node, rawCfg config.Node) error {
+	var (
+		insecureAuth bool
+		ioDebug      bool
+	)
+
+	cfg := config.Map{}
+	cfg.Custom("auth", false, false, defaultAuthProvider, authDirective, &endp.Auth)
+	cfg.Custom("storage", false, false, defaultStorage, storageDirective, &endp.Store)
+	cfg.Custom("tls", true, true, nil, tlsDirective, &endp.tlsConfig)
+	cfg.Bool("insecure_auth", false, &insecureAuth)
+	cfg.Bool("io_debug", false, &insecureAuth)
```

Comment:
`&ioDebug`?

Reply:
TBH I'm not sure we want to expose this as a configuration option

Reply:
Oops!

Reply:
Well, this is meant for debugging purposes. Perhaps we may want to ask users to enable it while providing logs in bug reports so we can know what clients do exactly.

Reply:
Okay, that's fair.

Reply:
(Keep in mind it'll leak passwords and sensitive data -- we should be careful about this. Maybe add a log line warning about it?)